io.trino.spi.TrinoException: Failed reading parquet data: Socket is closed by peer

After upgrading to 361, I'm facing an issue when running a fairly straight forward query:
```
SELECT * FROM some_table WHERE some_column = 'some_value' LIMIT 10
```
However, if I remove `WHERE` clause, it works as expected.

I have verified that parquet file is not corrupted in any way, and is indeed readable. Same query, using the same data source works as expected in v360.

Here's the full error:

```
io.trino.spi.TrinoException: Failed reading parquet data; source= s3://<REDACTED>; can not read class org.apache.parquet.format.PageHeader: Socket is closed by peer.
	at io.trino.plugin.hive.parquet.ParquetPageSource$ParquetBlockLoader.load(ParquetPageSource.java:230)
	at io.trino.spi.block.LazyBlock$LazyData.load(LazyBlock.java:396)
	at io.trino.spi.block.LazyBlock$LazyData.getFullyLoadedBlock(LazyBlock.java:375)
	at io.trino.spi.block.LazyBlock.getLoadedBlock(LazyBlock.java:282)
	at io.trino.operator.project.DictionaryAwarePageFilter.filter(DictionaryAwarePageFilter.java:59)
	at io.trino.operator.project.PageProcessor.createWorkProcessor(PageProcessor.java:121)
	at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.lambda$processPageSource$1(ScanFilterAndProjectOperator.java:293)
	at io.trino.operator.WorkProcessorUtils.lambda$flatMap$4(WorkProcessorUtils.java:245)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:319)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils.lambda$flatten$6(WorkProcessorUtils.java:277)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:319)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.trino.operator.WorkProcessorUtils.lambda$finishWhen$3(WorkProcessorUtils.java:215)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:151)
	at io.trino.operator.Driver.processInternal(Driver.java:387)
	at io.trino.operator.Driver.lambda$processFor$9(Driver.java:291)
	at io.trino.operator.Driver.tryWithLock(Driver.java:683)
	at io.trino.operator.Driver.processFor(Driver.java:284)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1076)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
	at io.trino.$gen.Trino_361____20210901_181740_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.IOException: can not read class org.apache.parquet.format.PageHeader: Socket is closed by peer.
	at org.apache.parquet.format.Util.read(Util.java:365)
	at org.apache.parquet.format.Util.readPageHeader(Util.java:132)
	at org.apache.parquet.format.Util.readPageHeader(Util.java:127)
	at io.trino.parquet.reader.ParquetColumnChunk.readPageHeader(ParquetColumnChunk.java:76)
	at io.trino.parquet.reader.ParquetColumnChunk.readAllPages(ParquetColumnChunk.java:89)
	at io.trino.parquet.reader.ParquetReader.createPageReader(ParquetReader.java:388)
	at io.trino.parquet.reader.ParquetReader.readPrimitive(ParquetReader.java:368)
	at io.trino.parquet.reader.ParquetReader.readColumnChunk(ParquetReader.java:444)
	at io.trino.parquet.reader.ParquetReader.readBlock(ParquetReader.java:427)
	at io.trino.plugin.hive.parquet.ParquetPageSource$ParquetBlockLoader.load(ParquetPageSource.java:224)
	... 39 more
Caused by: io.trino.hive.$internal.parquet.org.apache.thrift.transport.TTransportException: Socket is closed by peer.
	at io.trino.hive.$internal.parquet.org.apache.thrift.transport.TIOStreamTransport.read(TIOStreamTransport.java:130)
	at io.trino.hive.$internal.parquet.org.apache.thrift.transport.TTransport.readAll(TTransport.java:86)
	at io.trino.hive.$internal.parquet.org.apache.thrift.protocol.TCompactProtocol.readByte(TCompactProtocol.java:635)
	at io.trino.hive.$internal.parquet.org.apache.thrift.protocol.TCompactProtocol.readFieldBegin(TCompactProtocol.java:541)
	at org.apache.parquet.format.InterningProtocol.readFieldBegin(InterningProtocol.java:155)
	at org.apache.parquet.format.PageHeader$PageHeaderStandardScheme.read(PageHeader.java:1026)
	at org.apache.parquet.format.PageHeader$PageHeaderStandardScheme.read(PageHeader.java:1019)
	at org.apache.parquet.format.PageHeader.read(PageHeader.java:896)
	at org.apache.parquet.format.Util.read(Util.java:362)
	... 48 more
```

`SET SESSION hive.parquet_ignore_statistics = true` seems to bypass the issue. And the query works as expected on v361.

--- a/lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java
+++ b/lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java
@@ -127,8 +127,8 @@ class FilteredOffsetIndex
                 }
                 else {
                     ranges.add(new OffsetRange(currentOffset, currentLength));
-                    currentOffset = -1;
-                    currentLength = 0;
+                    currentOffset = offset;
+                    currentLength = length;
                 }
             }
 

--- a/lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java
+++ b/lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java
@@ -112,17 +112,13 @@ class FilteredOffsetIndex
                 ranges.add(new OffsetRange(rowGroupOffset, firstPageOffset - rowGroupOffset));
             }
 
-            long currentOffset = -1;
-            long currentLength = 0;
-            for (int i = 0; i < pageCount; ++i) {
+            long currentOffset = getOffset(0);
+            long currentLength = getCompressedPageSize(0);
+            for (int i = 1; i < pageCount; ++i) {
                 long offset = getOffset(i);
                 int length = getCompressedPageSize(i);
 
-                if (currentOffset == -1) {
-                    currentOffset = offset;
-                    currentLength = length;
-                }
-                else if (currentOffset + currentLength == offset) {
+                if (currentOffset + currentLength == offset) {
                     currentLength += length;
                 }
                 else {
@@ -131,10 +127,7 @@ class FilteredOffsetIndex
                     currentLength = length;
                 }
             }
-
-            if (currentOffset != -1) {
-                ranges.add(new OffsetRange(currentOffset, currentLength));
-            }
+            ranges.add(new OffsetRange(currentOffset, currentLength));
         }
         return ranges;
     }

--- a/lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetColumnChunk.java
+++ b/lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetColumnChunk.java
@@ -34,6 +34,7 @@ import java.util.List;
 import java.util.Optional;
 import java.util.OptionalLong;
 
+import static com.google.common.base.Verify.verify;
 import static io.trino.parquet.ParquetTypeUtils.getParquetEncoding;
 import static java.util.Objects.requireNonNull;
 
@@ -73,6 +74,7 @@ public class ParquetColumnChunk
     protected PageHeader readPageHeader()
             throws IOException
     {
+        verify(input.available() > 0, "Reached end of input unexpectedly");
         PageHeader pageHeader = Util.readPageHeader(input);
         advanceIfNecessary();
         return pageHeader;

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestParquetPageSkipping.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestParquetPageSkipping.java
@@ -86,7 +86,7 @@ public class TestParquetPageSkipping
     {
         String tableName = "test_and_predicate_" + randomTableSuffix();
         buildSortedTables(tableName, "totalprice", "double");
-        int rowCount = assertSameResults("SELECT * FROM " + tableName + " WHERE totalprice BETWEEN 100000 AND 131280 AND clerk = 'Clerk#000000624'");
+        int rowCount = assertColumnIndexResults("SELECT * FROM " + tableName + " WHERE totalprice BETWEEN 100000 AND 131280 AND clerk = 'Clerk#000000624'");
         assertThat(rowCount).isGreaterThan(0);
 
         // `totalprice BETWEEN 51890 AND 51900` is chosen to lie between min/max values of row group
@@ -105,20 +105,20 @@ public class TestParquetPageSkipping
             Object middleLowValue = values[1];
             Object middleHighValue = values[2];
             Object highValue = values[3];
-            assertSameResults(format("SELECT %s FROM %s WHERE %s = %s", sortByColumn, tableName, sortByColumn, middleLowValue));
-            assertThat(assertSameResults(format("SELECT %s FROM %s WHERE %s < %s ORDER BY %s", sortByColumn, tableName, sortByColumn, lowValue, sortByColumn))).isGreaterThan(0);
-            assertThat(assertSameResults(format("SELECT %s FROM %s WHERE %s > %s ORDER BY %s", sortByColumn, tableName, sortByColumn, highValue, sortByColumn))).isGreaterThan(0);
-            assertThat(assertSameResults(format("SELECT %s FROM %s WHERE %s BETWEEN %s AND %s ORDER BY %s", sortByColumn, tableName, sortByColumn, middleLowValue, middleHighValue, sortByColumn))).isGreaterThan(0);
+            assertColumnIndexResults(format("SELECT %s FROM %s WHERE %s = %s", sortByColumn, tableName, sortByColumn, middleLowValue));
+            assertThat(assertColumnIndexResults(format("SELECT %s FROM %s WHERE %s < %s ORDER BY %s", sortByColumn, tableName, sortByColumn, lowValue, sortByColumn))).isGreaterThan(0);
+            assertThat(assertColumnIndexResults(format("SELECT %s FROM %s WHERE %s > %s ORDER BY %s", sortByColumn, tableName, sortByColumn, highValue, sortByColumn))).isGreaterThan(0);
+            assertThat(assertColumnIndexResults(format("SELECT %s FROM %s WHERE %s BETWEEN %s AND %s ORDER BY %s", sortByColumn, tableName, sortByColumn, middleLowValue, middleHighValue, sortByColumn))).isGreaterThan(0);
             // Tests synchronization of reading values across columns
-            assertSameResults(format("SELECT * FROM %s WHERE %s = %s ORDER BY orderkey", tableName, sortByColumn, middleLowValue));
-            assertThat(assertSameResults(format("SELECT * FROM %s WHERE %s < %s ORDER BY orderkey", tableName, sortByColumn, lowValue))).isGreaterThan(0);
-            assertThat(assertSameResults(format("SELECT * FROM %s WHERE %s > %s ORDER BY orderkey", tableName, sortByColumn, highValue))).isGreaterThan(0);
-            assertThat(assertSameResults(format("SELECT * FROM %s WHERE %s BETWEEN %s AND %s ORDER BY orderkey", tableName, sortByColumn, middleLowValue, middleHighValue))).isGreaterThan(0);
+            assertColumnIndexResults(format("SELECT * FROM %s WHERE %s = %s ORDER BY orderkey", tableName, sortByColumn, middleLowValue));
+            assertThat(assertColumnIndexResults(format("SELECT * FROM %s WHERE %s < %s ORDER BY orderkey", tableName, sortByColumn, lowValue))).isGreaterThan(0);
+            assertThat(assertColumnIndexResults(format("SELECT * FROM %s WHERE %s > %s ORDER BY orderkey", tableName, sortByColumn, highValue))).isGreaterThan(0);
+            assertThat(assertColumnIndexResults(format("SELECT * FROM %s WHERE %s BETWEEN %s AND %s ORDER BY orderkey", tableName, sortByColumn, middleLowValue, middleHighValue))).isGreaterThan(0);
         }
         assertUpdate("DROP TABLE " + tableName);
     }
 
-    private int assertSameResults(String query)
+    private int assertColumnIndexResults(String query)
     {
         MaterializedResult withColumnIndexing = computeActual(query);
         MaterializedResult withoutColumnIndexing = computeActual(noParquetColumnIndexFiltering(getSession()), query);

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestParquetPageSkipping.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestParquetPageSkipping.java
@@ -23,7 +23,7 @@ import org.intellij.lang.annotations.Language;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
 
-import static io.trino.testing.assertions.Assert.assertEquals;
+import static io.trino.testing.QueryAssertions.assertEqualsIgnoreOrder;
 import static io.trino.testing.sql.TestTable.randomTableSuffix;
 import static java.lang.String.format;
 import static java.util.concurrent.TimeUnit.SECONDS;
@@ -122,7 +122,7 @@ public class TestParquetPageSkipping
     {
         MaterializedResult withColumnIndexing = computeActual(query);
         MaterializedResult withoutColumnIndexing = computeActual(noParquetColumnIndexFiltering(getSession()), query);
-        assertEquals(withColumnIndexing, withoutColumnIndexing);
+        assertEqualsIgnoreOrder(withColumnIndexing, withoutColumnIndexing);
         return withoutColumnIndexing.getRowCount();
     }
 

--- a/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestParquetPageSkipping.java
+++ b/plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestParquetPageSkipping.java
@@ -95,6 +95,29 @@ public class TestParquetPageSkipping
         assertUpdate("DROP TABLE " + tableName);
     }
 
+    @Test
+    public void testPageSkippingWithNonSequentialOffsets()
+    {
+        String tableName = "test_random_" + randomTableSuffix();
+        int updateCount = 8192;
+        assertUpdate(
+                "CREATE TABLE " + tableName + " (col) WITH (format = 'PARQUET') AS " +
+                        "SELECT * FROM unnest(transform(repeat(1, 8192), x -> rand()))",
+                updateCount);
+        for (int i = 0; i < 8; i++) {
+            assertUpdate(
+                    "INSERT INTO " + tableName + " SELECT rand() FROM " + tableName,
+                    updateCount);
+            updateCount += updateCount;
+        }
+        // These queries select a subset of pages which are stored at non-sequential offsets
+        // This reproduces the issue identified in https://github.com/trinodb/trino/issues/9097
+        for (double i = 0; i < 1; i += 0.1) {
+            assertColumnIndexResults(format("SELECT * FROM %s WHERE col BETWEEN %f AND %f", tableName, i - 0.00001, i + 0.00001));
+        }
+        assertUpdate("DROP TABLE " + tableName);
+    }
+
     @Test(dataProvider = "dataType")
     public void testPageSkipping(String sortByColumn, String sortByColumnType, Object[][] valuesArray)
     {

